<img src='https://encrypted-tbn0.gstatic.com/images?q=tbn%3AANd9GcQ-VfNtOyJbsaxu43Kztf_cv1mgBG6ZIQZEVw&usqp=CAU'>

# Procesamiento de Lenguaje Natural

## Taller #5: TF-IDF
`Estudiante: MARIA PAULA SILVA`



In [22]:
import re
import numpy as np
import pandas as pd

from nltk.corpus import stopwords

from nltk.stem.snowball import SnowballStemmer
spanishStemmer=SnowballStemmer("spanish")

from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.metrics.pairwise import cosine_distances

In [17]:
import nltk
nltk.download('all')

[nltk_data] Downloading collection 'all'
[nltk_data]    | 
[nltk_data]    | Downloading package abc to /root/nltk_data...
[nltk_data]    |   Unzipping corpora/abc.zip.
[nltk_data]    | Downloading package alpino to /root/nltk_data...
[nltk_data]    |   Unzipping corpora/alpino.zip.
[nltk_data]    | Downloading package biocreative_ppi to
[nltk_data]    |     /root/nltk_data...
[nltk_data]    |   Unzipping corpora/biocreative_ppi.zip.
[nltk_data]    | Downloading package brown to /root/nltk_data...
[nltk_data]    |   Unzipping corpora/brown.zip.
[nltk_data]    | Downloading package brown_tei to /root/nltk_data...
[nltk_data]    |   Unzipping corpora/brown_tei.zip.
[nltk_data]    | Downloading package cess_cat to /root/nltk_data...
[nltk_data]    |   Unzipping corpora/cess_cat.zip.
[nltk_data]    | Downloading package cess_esp to /root/nltk_data...
[nltk_data]    |   Unzipping corpora/cess_esp.zip.
[nltk_data]    | Downloading package chat80 to /root/nltk_data...
[nltk_data]    |   Unzipp

True

In [18]:
stopwords_sp = stopwords.words('spanish')

# Punto 1: Pre-Procesamiento

- `[9 pts]` Leer el archivo `bob_esponja.csv` usando `pandas`
- `[9 pts]` Crear una nueva columna con el texto en minúscula, sin caracteres especiales ni números, sin palabras vacias y hacer stemming de las palabras

In [10]:
#Leer datos
datos = pd.read_csv('bob_esponja.csv', sep = ",")
datos

Personaje                                        Descripción
0  Bob Esponja Pantalones Cuadrados  Bob Esponja es una esponja de mar con forma re...
1                 Patricio Estrella  Es una estrella de mar de color rosa, el mejor...
2              Calamardo Tentáculos  Calamardo Tentáculos es un personaje principal...
3                  Arenita Mejillas  Arenita Mejillas (Sandy Cheeks en inglés) en u...
4               Eugenio H. Cangrejo  Don Eugenio H. Cangrejo (normalmente llamado S...
5               Sheldon J. Plankton  Sheldon J. Plankton, o simplemente Plankton, e...
6                    Karen Plankton  Karen Plankton es uno de los dos principales a...
7                  Perlita Cangrejo  Perla "Perlita" Cangrejo es un personaje princ...
8                         Sra. Puff  La Señora Puff es un personaje principal de Bo...
9                   Gary el Caracol  Gary el Caracol es un personaje principal de B...

In [21]:
# Nueva columna
def pre_procesado(texto):
    texto = texto.lower()
    texto = re.sub(r"[\W\d_]+", " ", texto)
    texto = [palabra for palabra in texto.split() if palabra not in stopwords_sp]
    texto = " ".join(texto)
    return texto

datos['pre-procesado'] = datos['Descripción'].apply(lambda texto: pre_procesado(texto))

datos


Personaje  ...                                      pre-procesado
0  Bob Esponja Pantalones Cuadrados  ...  bob esponja esponja mar forma rectangular colo...
1                 Patricio Estrella  ...  estrella mar color rosa mejor amigo bob esponj...
2              Calamardo Tentáculos  ...  calamardo tentáculos personaje principal bob e...
3                  Arenita Mejillas  ...  arenita mejillas sandy cheeks inglés personaje...
4               Eugenio H. Cangrejo  ...  don eugenio h cangrejo normalmente llamado señ...
5               Sheldon J. Plankton  ...  sheldon j plankton simplemente plankton dos an...
6                    Karen Plankton  ...  karen plankton dos principales antagonistas bo...
7                  Perlita Cangrejo  ...  perla perlita cangrejo personaje principal bob...
8                         Sra. Puff  ...  señora puff personaje principal bob esponja ma...
9                   Gary el Caracol  ...  gary caracol personaje principal bob esponja q...

[10 rows x 3 columns]

# Punto 2: TF-IDF

- `[16 pts]` Crear la matriz TF-IDF

In [ ]:
datos['pre-procesado'].values

In [30]:
tfidf_vector = TfidfVectorizer()
tfidf = tfidf_vector.fit_transform(datos['pre-procesado'].values)

tfidf.toarray()




array([[0.05870539, 0.        , 0.        , ..., 0.        , 0.05870539,
        0.        ],
       [0.        , 0.        , 0.        , ..., 0.        , 0.        ,
        0.        ],
       [0.        , 0.09071203, 0.09071203, ..., 0.        , 0.        ,
        0.        ],
       ...,
       [0.        , 0.        , 0.        , ..., 0.        , 0.        ,
        0.        ],
       [0.        , 0.        , 0.        , ..., 0.        , 0.        ,
        0.        ],
       [0.        , 0.        , 0.        , ..., 0.09289037, 0.        ,
        0.        ]])

In [35]:
tfidf_matrix = pd.DataFrame(tfidf.toarray())
tfidf_matrix.columns = tfidf_vector.get_feature_names()
tfidf_matrix.index = datos.index

tfidf_matrix.T

0    1         2  ...    7         8         9
abajo       0.058705  0.0  0.000000  ...  0.0  0.000000  0.000000
aborrece    0.000000  0.0  0.090712  ...  0.0  0.000000  0.000000
acaba       0.000000  0.0  0.090712  ...  0.0  0.000000  0.000000
accidente   0.049905  0.0  0.000000  ...  0.0  0.000000  0.000000
aceptó      0.000000  0.0  0.000000  ...  0.0  0.000000  0.000000
...              ...  ...       ...  ...  ...       ...       ...
zapatos     0.043661  0.0  0.000000  ...  0.0  0.091678  0.069085
árbol       0.000000  0.0  0.000000  ...  0.0  0.000000  0.000000
éste        0.000000  0.0  0.000000  ...  0.0  0.000000  0.092890
única       0.058705  0.0  0.000000  ...  0.0  0.000000  0.000000
únicamente  0.000000  0.0  0.000000  ...  0.0  0.000000  0.000000

[835 rows x 10 columns]

# Punto 3: Distancia del coseno
- `[12 pts]` Calcular la distancia del coseno entre cada una de los personajes
- `[2 pts]` ¿Cuáles son los personajes más parecidos?
- `[2 pts]` ¿Cuáles son los personajes más diferentes?

In [39]:
dist_cos = cosine_distances(tfidf_matrix.values)
dist_cos = pd.DataFrame(dist_cos, columns = tfidf_matrix.index, index = tfidf_matrix.index)
dist_cos

0         1         2  ...         7         8         9
0  0.000000  0.850507  0.927257  ...  0.916038  0.882641  0.931071
1  0.850507  0.000000  0.808538  ...  0.936298  0.954744  0.922686
2  0.927257  0.808538  0.000000  ...  0.972910  0.968216  0.914684
3  0.871921  0.893063  0.943984  ...  0.982508  0.983039  0.930316
4  0.881220  0.893730  0.873121  ...  0.829929  0.938435  0.928366
5  0.901743  0.968058  0.921201  ...  0.921811  0.990016  0.971474
6  0.957920  0.961623  0.969450  ...  0.984115  0.991454  0.978322
7  0.916038  0.936298  0.972910  ...  0.000000  0.895400  0.967919
8  0.882641  0.954744  0.968216  ...  0.895400  0.000000  0.963591
9  0.931071  0.922686  0.914684  ...  0.967919  0.963591  0.000000

[10 rows x 10 columns]

# ¿Cuáles son los personajes más parecidos?

In [41]:
Menos_parecidos = dist_cos[dist_cos == np.amax(dist_cos.values)] # Obtener el mayor valor en la matriz
Menos_parecidos

0   1   2   3   4   5         6   7         8   9
0 NaN NaN NaN NaN NaN NaN       NaN NaN       NaN NaN
1 NaN NaN NaN NaN NaN NaN       NaN NaN       NaN NaN
2 NaN NaN NaN NaN NaN NaN       NaN NaN       NaN NaN
3 NaN NaN NaN NaN NaN NaN       NaN NaN       NaN NaN
4 NaN NaN NaN NaN NaN NaN       NaN NaN       NaN NaN
5 NaN NaN NaN NaN NaN NaN       NaN NaN       NaN NaN
6 NaN NaN NaN NaN NaN NaN       NaN NaN  0.991454 NaN
7 NaN NaN NaN NaN NaN NaN       NaN NaN       NaN NaN
8 NaN NaN NaN NaN NaN NaN  0.991454 NaN       NaN NaN
9 NaN NaN NaN NaN NaN NaN       NaN NaN       NaN NaN

Los personajes menos parecidos son la Sra. Puff y Sheldon J. Plankton con un valor de 0.991454

# ¿Cuáles son los personajes más diferentes?

In [43]:
dist_cos.values[dist_cos.values==0]=2
Mas_parecidos = dist_cos[dist_cos == np.amin(dist_cos.values)]# Buscar el valor menor en la matriz
Mas_parecidos

0   1   2   3         4         5   6   7   8   9
0 NaN NaN NaN NaN       NaN       NaN NaN NaN NaN NaN
1 NaN NaN NaN NaN       NaN       NaN NaN NaN NaN NaN
2 NaN NaN NaN NaN       NaN       NaN NaN NaN NaN NaN
3 NaN NaN NaN NaN       NaN       NaN NaN NaN NaN NaN
4 NaN NaN NaN NaN       NaN  0.704685 NaN NaN NaN NaN
5 NaN NaN NaN NaN  0.704685       NaN NaN NaN NaN NaN
6 NaN NaN NaN NaN       NaN       NaN NaN NaN NaN NaN
7 NaN NaN NaN NaN       NaN       NaN NaN NaN NaN NaN
8 NaN NaN NaN NaN       NaN       NaN NaN NaN NaN NaN
9 NaN NaN NaN NaN       NaN       NaN NaN NaN NaN NaN

Los personajes más parecidos son Eugenio H. Cangrejo y Sheldon J. Plankton con un valor de 0.704685